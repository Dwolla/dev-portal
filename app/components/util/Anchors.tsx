import {
  Children,
  cloneElement,
  createContext,
  useContext,
  useEffect,
  useState,
} from "react";
import { useRouter } from "next/router";
import GithubSlugger from "github-slugger";
import scrollToElement from "scroll-to-element";
import { childrenToString } from "../../modules/helpers";

// constants

const ANCHOR_EL_TYPES = ["h1", "h2", "h3"];
export const SCROLL_OFFSET = 100;
export const SCROLL_DURATION = 500;

// AnchorsContext

export const AnchorsContext = createContext({
  anchors: [] as Anchor[],
  setAnchors: (_anchors: Anchor[]) => {}, // eslint-disable-line no-unused-vars
  activeAnchor: null as Anchor,
  setActiveAnchor: (anchor: Anchor) => {}, // eslint-disable-line no-unused-vars
});

export const useAnchors = () => useContext(AnchorsContext);

// AnchorsProvider

export function AnchorsProvider(props: {
  children: JSX.Element | JSX.Element[];
}) {
  const [anchors, setAnchors] = useState([]);
  const [activeAnchor, setActiveAnchor] = useState(null);

  return (
    <AnchorsContext.Provider
      value={{
        anchors,
        setAnchors,
        activeAnchor,
        setActiveAnchor,
      }}
    >
      {props.children}
    </AnchorsContext.Provider>
  );
}

// AnchorsSetter

function AnchorData(props: { anchor: Anchor; children: JSX.Element }) {
  return props.children;
}

export function AnchorsSetter(props) {
  const { setAnchors, activeAnchor, setActiveAnchor } = useAnchors();

  const slugger = new GithubSlugger();

  const children =
    Children.map(props.children, (c) => {
      if (ANCHOR_EL_TYPES.includes(c.props.originalType)) {
        const text = childrenToString(c.props.children);
        const id = slugger.slug(text);
        const level = ANCHOR_EL_TYPES.indexOf(c.props.originalType) + 1;
        return (
          <AnchorData anchor={{ id, text, level }}>
            {cloneElement(c, { id })}
          </AnchorData>
        );
      }
      return c;
    }) || [];

  const anchors = children.reduce(
    (acc, current) =>
      current.props.anchor ? [...acc, current.props.anchor] : acc,
    []
  );

  const router = useRouter();

  useEffect(() => {
    setAnchors(anchors);
    setActiveAnchor(anchors[0]);

    return () => {
      setAnchors([]);
      setActiveAnchor(null);
    };
  }, [router?.pathname]);

  useEffect(() => {
    let raf: number;
    const anchorEls = anchors.map((a: Anchor) => document.getElementById(a.id));

    const handleScroll = () => {
      raf = requestAnimationFrame(() => {
        if (anchorEls && anchorEls.length > 0) {
          const anchorId = anchorEls.reduce(
            (acc: { score: number; el: { id: string } }, next) => {
              const { y } = next.getBoundingClientRect();
              const nextScore = Math.abs(SCROLL_OFFSET - y);

              if (!acc) {
                return { el: next, score: nextScore };
              }

              return nextScore < acc.score
                ? { el: next, score: nextScore }
                : acc;
            },
            null
          ).el.id;

          if (anchorId !== activeAnchor?.id) {
            setActiveAnchor(anchors.find((a: Anchor) => a.id === anchorId));
          }
        }
      });
    };

    window.addEventListener("scroll", handleScroll);
    window.addEventListener("resize", handleScroll);

    return () => {
      window.removeEventListener("scroll", handleScroll);
      window.removeEventListener("resize", handleScroll);
      cancelAnimationFrame(raf);
    };
  }, [router?.pathname]);

  useEffect(() => {
    if (activeAnchor) {
      window.history.replaceState(
        null,
        null,
        activeAnchor.level !== 1 ? `#${activeAnchor.id}` : " " //prevent h1 tags from displaying as anchors in the url
      );
    }
  }, [activeAnchor?.id]);

  return children;
}

export const scrollTo = (anchorId: string) => {
  const element = document.getElementById(anchorId);

  if (element) {
    scrollToElement(element, {
      offset: -SCROLL_OFFSET + 1,
      duration: SCROLL_DURATION,
    });
  }
};
