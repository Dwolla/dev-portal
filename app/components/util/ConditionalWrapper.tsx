// Used for conditionally wrapping elements

const ConditionalWrapper = ({ condition, wrapper, children }) =>
  condition ? wrapper(children) : children;

export default ConditionalWrapper;
