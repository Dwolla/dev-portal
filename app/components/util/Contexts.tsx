import React from "react";

// eslint-disable-next-line import/prefer-default-export
export const LanguageContext = React.createContext({
  selectedLanguage: { value: "", label: "" },
  setSelectedLanguage: (language) => {
    return language;
  },
  languageOptions: [],
});
