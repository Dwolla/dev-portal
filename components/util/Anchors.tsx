import {
  createContext,
  useState,
  useEffect,
  useContext,
  Children,
  cloneElement,
} from "react";
import { useRouter } from "next/router";
import GithubSlugger from "github-slugger";
import { childrenToString } from "../../modules/helpers";

// constants

const ANCHOR_EL_TYPES = ["h1", "h2", "h3", "h4", "h5", "h6"];

// AnchorsContext

export const AnchorsContext = createContext({
  anchors: [] as Anchor[],
  setAnchors: (_anchors: Anchor[]) => {}, // eslint-disable-line no-unused-vars
});

export const useAnchors = () => useContext(AnchorsContext);

// AnchorsProvider

export function AnchorsProvider(props: {
  children: JSX.Element | JSX.Element[];
}) {
  const [anchors, setAnchors] = useState([]);

  return (
    <AnchorsContext.Provider value={{ anchors, setAnchors }}>
      {props.children}
    </AnchorsContext.Provider>
  );
}

// AnchorsSetter

function AnchorData(props: { anchor: Anchor; children: JSX.Element }) {
  return props.children;
}

export function AnchorsSetter(props) {
  const { setAnchors } = useAnchors();

  const slugger = new GithubSlugger();

  const children = Children.map(props.children, (c) => {
    if (ANCHOR_EL_TYPES.includes(c.props.originalType)) {
      const text = childrenToString(c.props.children);
      const id = slugger.slug(text);
      const level = ANCHOR_EL_TYPES.indexOf(c.props.originalType) + 1;
      return (
        <AnchorData anchor={{ id, text, level }}>
          {cloneElement(c, { id })}
        </AnchorData>
      );
    }
    return c;
  });

  const anchors = children.reduce(
    (acc, current) =>
      current.props.anchor ? [...acc, current.props.anchor] : acc,
    []
  );

  const router = useRouter();

  useEffect(() => {
    setAnchors(anchors);

    return () => setAnchors([]);
  }, [router?.pathname]);

  return children;
}
