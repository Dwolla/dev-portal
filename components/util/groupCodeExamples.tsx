/* eslint-disable react/no-array-index-key */
import { Children, cloneElement } from "react";

const getCodeBlock = (el) =>
  el.props.originalType === "pre" &&
  Children.only(el.props.children).props.originalType === "code"
    ? Children.only(el.props.children)
    : null;

export const getLanguage = (el) =>
  getCodeBlock(el).props.className?.replace("language-", "") || "plaintext";

export default function groupCodeExamples({
  into: Component,
  children,
}: {
  into: any;
  children: any;
}): JSX.Element {
  const grouped = Children.toArray(children).reduce(
    (acc: any, next: any, i) => {
      if (acc === null) {
        return [
          getCodeBlock(next) ? (
            <Component key={i}>{next}</Component>
          ) : (
            cloneElement(next, { key: i })
          ),
        ];
      }

      const [head, last] = [acc.slice(0, acc.length - 1), acc[acc.length - 1]];

      const tail =
        last.type === Component && getCodeBlock(next)
          ? [
              cloneElement(last, {
                children: [...Children.toArray(last.props.children), next],
              }),
            ]
          : [
              last,
              getCodeBlock(next) ? (
                <Component key={i}>{next}</Component>
              ) : (
                cloneElement(next, { key: i })
              ),
            ];

      return [...head, ...tail];
    },
    null
  );

  return grouped as JSX.Element;
}
