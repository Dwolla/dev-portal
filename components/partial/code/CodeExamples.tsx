import { Children, useContext, useMemo, useState, useEffect } from "react";
import sortBy from "lodash.sortby";
import { disableBodyScroll, enableBodyScroll } from "body-scroll-lock";
import { getLanguage } from "../../util/groupCodeExamples";
import { LanguageContext } from "../../util/Contexts";
// eslint-disable-next-line no-unused-vars
import highlight, { Language } from "../../../modules/highlight";
import Select from "../../base/select/Select";
import {
  ExpandedOverlay,
  ExpandedCodeBlock,
  CodeBlockBar,
  BarText,
  CodeBlockButton,
  CodeBlockContent,
} from "./CodeExamples.styled";
import { ReactComponent as CopyIcon } from "../../../assets/images/component-icons/copy-icon.svg";
import { ReactComponent as ExpandIcon } from "../../../assets/images/component-icons/expand.svg";
import { ReactComponent as CollapseIcon } from "../../../assets/images/component-icons/collapse.svg";
import useCopy from "../../../hooks/useCopy";
import ga from "../../../modules/ga";

const getInitialLanguage = (
  exampleLanguages: string[],
  ctxSelectedLanguage: string
) => {
  const matchingCtxSelectedLanguage = exampleLanguages.find(
    (l) => l === ctxSelectedLanguage
  );
  return matchingCtxSelectedLanguage || exampleLanguages[0];
};

const getLanguageOptions = (language: string, ctxLanguageOptions) =>
  ctxLanguageOptions.find((lo) => lo.value === language);

const codeStringFrom = (codeExample: any): any =>
  Children.only(codeExample.props.children).props.children;

type CodeExampleProps = { language: Language; children: string };

// eslint-disable-next-line no-unused-vars
export const CodeExample = (props: CodeExampleProps) => null;

const toCodeExampleProps = (child): CodeExampleProps => {
  if (child.type === CodeExample || child.props.mdxType === "CodeExample") {
    return child.props;
  }
  return { language: getLanguage(child), children: codeStringFrom(child) };
};

export default function CodeExamples({
  children,
}: {
  children: JSX.Element | JSX.Element[];
}) {
  const examples: CodeExampleProps[] = Children.toArray(children).map(
    toCodeExampleProps
  );

  const ctx = useContext(LanguageContext);

  const exampleLanguages: string[] = sortBy(
    examples.map((e) => e.language),
    (l) => ctx.languageOptions.findIndex(({ value }) => value === l)
  );

  const [selectedLanguage, setSelectedLanguage] = useState(null);
  const [expanded, setExpanded] = useState("");

  const activeLanguage =
    selectedLanguage ||
    getInitialLanguage(exampleLanguages, ctx.selectedLanguage.value);

  // when ctx.selectedLanguage changes, setSelectedLanguage if
  // ctx.selectedLanguage is one of the exampleLanguages
  useEffect(() => {
    if (exampleLanguages.includes(ctx.selectedLanguage.value)) {
      setSelectedLanguage(ctx.selectedLanguage.value);
    }
  }, [ctx.selectedLanguage.value]);

  // when selectedLanguage changes, update the context if
  // selectedLanguage is one of the context's languageOptions
  useEffect(() => {
    const selectedLanguageOptionsFromCtx = getLanguageOptions(
      selectedLanguage,
      ctx.languageOptions
    );
    if (selectedLanguageOptionsFromCtx) {
      ctx.setSelectedLanguage(selectedLanguageOptionsFromCtx);
    }
  }, [selectedLanguage]);

  const { children: codeString, language: codeLanguage } =
    examples.find((e) => e.language === activeLanguage) || examples[0];

  const codeHtmlHighlighted = useMemo(
    () => highlight(codeString, codeLanguage),
    [codeString, codeLanguage]
  );

  const options = exampleLanguages.map(
    (l) => getLanguageOptions(l, ctx.languageOptions) || { value: l, label: l }
  );

  const { copied, copy } = useCopy(codeString);

  const onlyOneExample = options.length === 1;

  const toggleExpanded = () => {
    if (!expanded) {
      setExpanded("expanded");
      disableBodyScroll();
    } else {
      setExpanded("collapsing");
      setTimeout(() => {
        setExpanded("");
        enableBodyScroll();
      }, 280);
    }
  };

  type CodeBlockProps = {
    variant?: "default" | "fullscreen";
  };

  const CodeBlock = ({ variant = "default" }: CodeBlockProps) => (
    <>
      <CodeBlockBar className={`code-block-bar-${variant}`}>
        {onlyOneExample ? (
          <BarText>{activeLanguage}</BarText>
        ) : (
          <Select
            autoWidth
            variant="code"
            options={options}
            selectedValue={options.find((o) => o.value === activeLanguage)}
            setSelectedValue={({ value }) => {
              setSelectedLanguage(value);
              ga("code sample", "language change", value);
            }}
          />
        )}

        <CodeBlockButton type="button" onClick={copy} disabled={copied}>
          {copied ? "copied" : "copy"} <CopyIcon />
        </CodeBlockButton>

        <CodeBlockButton
          type="button"
          onClick={toggleExpanded}
          autoFocus={variant === "fullscreen"}
        >
          {variant === "fullscreen" ? (
            <>
              collapse <CollapseIcon width={16} height={16} />
            </>
          ) : (
            <>
              expand <ExpandIcon width={16} height={16} />
            </>
          )}
        </CodeBlockButton>
      </CodeBlockBar>

      <CodeBlockContent className={`code-block-${variant}`}>
        <pre>
          <code
            className={`language-${activeLanguage}`}
            // eslint-disable-next-line react/no-danger
            dangerouslySetInnerHTML={{
              __html: codeHtmlHighlighted,
            }}
          />
        </pre>
      </CodeBlockContent>
    </>
  );

  return (
    <div>
      <>
        <CodeBlock />
      </>

      {expanded && (
        <ExpandedOverlay className={expanded}>
          <ExpandedCodeBlock className={expanded}>
            <CodeBlock variant="fullscreen" />
          </ExpandedCodeBlock>
        </ExpandedOverlay>
      )}
    </div>
  );
}
