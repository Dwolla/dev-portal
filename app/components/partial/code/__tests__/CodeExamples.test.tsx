/* eslint-disable react/jsx-curly-brace-presence */
import renderer from "react-test-renderer";
import { createElement } from "react";
import CodeExamples, { CodeExample } from "../CodeExamples";
import { LanguageContext } from "../../../util/Contexts";

function MockMDXCreateElement(props: {
  children: any;
  originalType: string;
  id?: string;
  className?: string;
}) {
  return createElement(props.originalType, {
    children: props.children,
    id: props.id,
  });
}

test("single example", () => {
  const tree = renderer
    .create(
      <LanguageContext.Provider
        value={{
          selectedLanguage: { value: "php", label: "PHP" },
          setSelectedLanguage: () => {},
          languageOptions: [
            { value: "java", label: "Java" },
            { value: "php", label: "PHP" },
          ],
        }}
      >
        <CodeExamples>
          <MockMDXCreateElement originalType="pre">
            <MockMDXCreateElement
              originalType="code"
              className="language-javascript"
            >
              {'const foo = "bar";'}
            </MockMDXCreateElement>
          </MockMDXCreateElement>
        </CodeExamples>
      </LanguageContext.Provider>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

test("multiple examples", () => {
  const tree = renderer
    .create(
      <LanguageContext.Provider
        value={{
          selectedLanguage: { value: "php", label: "PHP" },
          setSelectedLanguage: () => {},
          languageOptions: [
            { value: "java", label: "Java" },
            { value: "php", label: "PHP" },
          ],
        }}
      >
        <CodeExamples>
          <CodeExample language="javascript">
            {'const foo = "bar";'}
          </CodeExample>

          <MockMDXCreateElement originalType="pre">
            <MockMDXCreateElement originalType="code">
              {"plaintext goes here"}
            </MockMDXCreateElement>
          </MockMDXCreateElement>

          <MockMDXCreateElement originalType="pre">
            <MockMDXCreateElement originalType="code" className="language-ruby">
              {'foo = "bar"'}
            </MockMDXCreateElement>
          </MockMDXCreateElement>

          <MockMDXCreateElement originalType="pre">
            <MockMDXCreateElement originalType="code" className="language-php">
              {'$foo = "bar"'}
            </MockMDXCreateElement>
          </MockMDXCreateElement>

          <MockMDXCreateElement originalType="pre">
            <MockMDXCreateElement originalType="code" className="language-java">
              {'string foo = "bar"'}
            </MockMDXCreateElement>
          </MockMDXCreateElement>
        </CodeExamples>
      </LanguageContext.Provider>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
