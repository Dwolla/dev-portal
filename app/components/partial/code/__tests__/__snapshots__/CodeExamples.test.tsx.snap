// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`multiple examples 1`] = `
<div>
  <div
    className="code-block-bar-default  css-vlw6pb-CodeBlockBar e112hlxu2"
  >
    <div
      className="css-dffxm8-BarItems e112hlxu4"
    >
      <div
        className=" css-k377xg-container"
        onKeyDown={[Function]}
      >
        <div
          className=" css-ypb917-control"
          onMouseDown={[Function]}
          onTouchEnd={[Function]}
        >
          <div
            className=" css-1fpxn2j-ValueContainer"
          >
            <div
              className=" css-gn3biy-singleValue"
            >
              PHP
            </div>
            <input
              aria-autocomplete="list"
              className="css-62g3xt-dummyInput"
              disabled={false}
              id="react-select-2-input"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              readOnly={true}
              tabIndex="0"
              value=""
            />
          </div>
          <div
            className=" css-1hb7zxy-IndicatorsContainer"
          >
            <span
              className=" css-43ykx9-indicatorSeparator"
            />
            <div
              aria-hidden="true"
              className=" css-1hkylt5-indicatorContainer"
              onMouseDown={[Function]}
              onTouchEnd={[Function]}
            >
              <svg
                height={12}
                width={7}
              />
            </div>
          </div>
        </div>
      </div>
      <button
        className="css-5uu5vc-CodeBlockButton e112hlxu5"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        copy
         
        <svg />
      </button>
      <button
        autoFocus={false}
        className="css-5uu5vc-CodeBlockButton e112hlxu5"
        onClick={[Function]}
        type="button"
      >
        expand 
        <svg
          height={16}
          width={16}
        />
      </button>
    </div>
  </div>
  <div
    className="code-block-default css-1wvz7mj-CodeBlockContent e112hlxu7"
  >
    <pre>
      <code
        className="language-php"
        dangerouslySetInnerHTML={
          Object {
            "__html": "$foo = <span class=\\"hljs-string\\">\\"bar\\"</span>",
          }
        }
      />
    </pre>
  </div>
  
</div>
`;

exports[`single example 1`] = `
<div>
  <div
    className="code-block-bar-default  css-vlw6pb-CodeBlockBar e112hlxu2"
  >
    <div
      className="css-dffxm8-BarItems e112hlxu4"
    >
      <div
        className="css-1vovl9e-BarText e112hlxu6"
      >
        javascript
      </div>
      <button
        className="css-5uu5vc-CodeBlockButton e112hlxu5"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        copy
         
        <svg />
      </button>
      <button
        autoFocus={false}
        className="css-5uu5vc-CodeBlockButton e112hlxu5"
        onClick={[Function]}
        type="button"
      >
        expand 
        <svg
          height={16}
          width={16}
        />
      </button>
    </div>
  </div>
  <div
    className="code-block-default css-1wvz7mj-CodeBlockContent e112hlxu7"
  >
    <pre>
      <code
        className="language-javascript"
        dangerouslySetInnerHTML={
          Object {
            "__html": "<span class=\\"hljs-keyword\\">const</span> foo = <span class=\\"hljs-string\\">\\"bar\\"</span>;",
          }
        }
      />
    </pre>
  </div>
  
</div>
`;

exports[`single example with title 1`] = `
<div>
  <div
    className="code-block-bar-default title css-vlw6pb-CodeBlockBar e112hlxu2"
  >
    <div
      className="css-1v0t4mu-BarTitle e112hlxu3"
    >
      Title goes here
    </div>
    <div
      className="css-dffxm8-BarItems e112hlxu4"
    >
      <div
        className="css-1vovl9e-BarText e112hlxu6"
      >
        javascript
      </div>
      <button
        className="css-5uu5vc-CodeBlockButton e112hlxu5"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        copy
         
        <svg />
      </button>
      <button
        autoFocus={false}
        className="css-5uu5vc-CodeBlockButton e112hlxu5"
        onClick={[Function]}
        type="button"
      >
        expand 
        <svg
          height={16}
          width={16}
        />
      </button>
    </div>
  </div>
  <div
    className="code-block-default css-1wvz7mj-CodeBlockContent e112hlxu7"
  >
    <pre>
      <code
        className="language-javascript"
        dangerouslySetInnerHTML={
          Object {
            "__html": "<span class=\\"hljs-keyword\\">const</span> foo = <span class=\\"hljs-string\\">\\"bar\\"</span>;",
          }
        }
      />
    </pre>
  </div>
  
</div>
`;
